ffi/src/lib.rs

```diff
@@ -1,173 +1,381 @@
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-// NOTE: Enum is `repr(i32)` becasuse WebAssembly supports only
-// u32/i32, u64/i64 natively. Otherwise, `repr(i8)` would suffice
-#[repr(i32)]
+#[repr(i8)]
 pub enum FfiResult {
-    /// Handle id doesn't identify any known handles
-    UnknownHandle = -4_i32,
-    /// Executing the wrapped method on handle returned error
-    ExecutionFail = -3_i32,
-    /// Raw pointer input argument to FFI function was null
-    ArgIsNull = -2_i32,
-    /// Given bytes don't comprise a valid UTF8 string
-    Utf8Error = -1_i32,
-    /// FFI function executed successfully
-    Ok = 0_i32,
-}
-
-/// Implement `Handle` for given types with first argument as the initial handle id.
-#[macro_export]
-macro_rules! handles {
-    ( $id:expr, $ty:ty $(, $other:ty)* $(,)? ) => {
-        impl Handle for $ty {
-            const ID: $crate::HandleId = $id;
+    /// The input argument provided to FFI function has a trap representation.
+    TrapRepresentation = -6,
+    /// FFI function execution panicked.
+    UnrecoverableError = -5,
+    /// Provided handle id doesn't match any known handles.
+    UnknownHandle = -4,
+    /// FFI function failed during the execution of the wrapped method on the provided handle.
+    ExecutionFail = -3,
+    /// The input argument provided to FFI function is a null pointer.
+    ArgIsNull = -2,
+    /// The input argument provided to FFI function is not a valid UTF-8 string.
+    Utf8Error = -1,
+    /// FFI function executed successfully.
+    Ok = 0,
+}
+
+unsafe impl<T> ReprC for *const T {}
+unsafe impl<T> ReprC for *mut T {}
+
+impl<'itm, T: ReprC + Copy + 'itm> AsReprCRef<'itm> for T
+where
+    T: IntoFfi<Target = Self>,
+{
+    type Target = Self;
+
+    fn as_ref(&self) -> Self::Target {
+        *self
+    }
+}
+impl<'itm, T: 'itm> AsReprCRef<'itm> for *const T {
+    type Target = Self;
+
+    fn as_ref(&self) -> Self::Target {
+        *self
+    }
+}
+
+impl<T: ReprC + Copy> IntoFfi for &T
+where
+    T: IntoFfi<Target = T>,
+{
+    type Target = *const T;
+
+    fn into_ffi(self) -> Self::Target {
+        Self::Target::from(self)
+    }
+}
+impl<T: ReprC + Copy> IntoFfi for &mut T
+where
+    T: IntoFfi<Target = T>,
+{
+    type Target = *mut T;
+
+    fn into_ffi(self) -> Self::Target {
+        Self::Target::from(self)
+    }
+}
+
+impl<'itm, T: ReprC> TryFromReprC<'itm> for &'itm T {
+    type Source = *const T;
+    type Store = ();
+
+    unsafe fn try_from_repr_c(source: Self::Source, _: &mut ()) -> Result<Self, FfiResult> {
+        source.as_ref().ok_or(FfiResult::ArgIsNull)
+    }
+}
+impl<'itm, T: ReprC> TryFromReprC<'itm> for &'itm mut T {
+    type Source = *mut T;
+    type Store = ();
+
+    unsafe fn try_from_repr_c(source: Self::Source, _: &mut ()) -> Result<Self, FfiResult> {
+        source.as_mut().ok_or(FfiResult::ArgIsNull)
+    }
+}
+
+impl<T: ReprC> OutPtr for *mut T {
+    fn is_valid(&self) -> bool {
+        (*self).is_null()
```

Comment:
I think should be !(*self).is_null(), no?

Reply:
Yes